subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/IncrementalCompilerDecorator.java

```diff
@@ -83,4 +95,52 @@ public IncrementalCompilerDecorator(JarClasspathSnapshotMaker jarClasspathSnapsh
         PreviousCompilation previousCompilation = new PreviousCompilation(new ClassSetAnalysis(data), compileCaches.getLocalJarClasspathSnapshotStore(), compileCaches.getJarSnapshotCache());
         return new SelectiveCompiler(inputs, previousCompilation, cleaningCompiler, staleClassDetecter, compilationInitializer, jarClasspathSnapshotMaker);
     }
+
+    // Check whether there are any Annotation Processors on the classpath that are NOT
+    // incap-enabled (i.e. incremental).  If so, then we do not allow an incremental build.
+    // If there is no annotation processing at all, then we permit an incremental build.
+    private boolean allowIncrementalAnnotationProcessing() {
+        if (!compileOptions.isIncrementalAnnotationProcessing()) {
+            return false;
+        }
+        FileCollection classPath = new SimpleFileCollection(javaCompileSpec.getCompileClasspath());
+        Set<AnnotationProcessorInfo> cachedInfo = annotationProcessorDetector.getAnnotationProcessorInfo(compileOptions, classPath);
```

Comment:
The detector can probably do with `Iterable<File>` instead, which would avoid this wrapping.